src/main/java/org/pvv/shufflegene/Traverse.java

```diff
@@ -0,0 +1,163 @@
+ */
+public class Traverse {
+    public char start;
+    public char end;
+    // start, end and path of original sequence
+    public int length;
+    private HashMap<Character, ArrayList<Character>> edgeMap;
+
+    /**
+     * Construct a traverse from a sequence of nucleotides
+     *
+     * @param sequence string of nucleotides
+     */
+    public Traverse(String sequence) {
+        sequence = sequence.toUpperCase();
+        setEdgeMap(sequence);
+    }
+
+    public Traverse(char start, char end, int length, HashMap<Character, ArrayList<Character>> edgeMap) {
+        this.start = start;
+        this.end = end;
+        this.length = length;
+        this.edgeMap = edgeMap;
+    }
+
+    // Helper function to copy a list.
+    private static <T> ArrayList<T> copyList(List<T> input) {
+        ArrayList<T> result = new ArrayList<>(input);
+        Collections.copy(input, result);
+        return result;
+    }
+
+    /**
+     * Set the edge map and associated variables based on an input string.
+     *
+     * @param sequence nucleotide sequence.
+     */
+    private void setEdgeMap(String sequence) {
+        char[] nucleotides = sequence.toCharArray();
+        this.start = nucleotides[0];
+        this.end = nucleotides[nucleotides.length - 1];
+        this.length = nucleotides.length;
+
+        this.edgeMap = new HashMap<>();
+
+        for (int i = 0; i < nucleotides.length - 1; i++) {
+            ArrayList<Character> edges = this.edgeMap.getOrDefault(nucleotides[i], new ArrayList<>());
+            edges.add(nucleotides[i + 1]);
+            this.edgeMap.put(nucleotides[i], edges);
+        }
+
+    }
+
+    /**
+     * Remove the first occurrence of a given edge from the traverse.
+     *
+     * @param edge Pair of characters giving the start and end of the vertex.
+     * @return whether the removal was successful or not.
+     */
+    public boolean removeEdge(Pair<Character, Character> edge) {
+        return edgeMap.get(edge.getValue0()).remove(edge.getValue1());
+    }
+
+    /**
+     * Append an edge to the end of the traverse.
+     *
+     * @param edge Pair of characters giving the start and end of the vertex.
+     */
+    public void appendEdge(Pair<Character, Character> edge) {
+        edgeMap.get(edge.getValue0()).add(edge.getValue1());
+    }
+
+    /**
+     * Perform a deep copy of edgeMap, used by {@link #deepCopy()}
+     *
+     * @return a copy of edgeMap
+     */
+    private HashMap<Character, ArrayList<Character>> copyEdgeMap() {
+        return this.edgeMap
+                .entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> copyList(e.getValue()),
+                        (v1, v2) -> {
+                            throw new IllegalStateException("Duplicate key");
+                        },
+                        HashMap::new
+                ));
```

Comment:
Mm. I'm also not sure this one is needed since I know create one of these objects for each shuffle. I guess the root problem is the destructive unsafeToString method. Perhaps I'll actually go the extra mile to avoid popping elements off the list.